### PR TITLE
Arc 1533 conditional proxy

### PR DESCRIPTION
**Pull request by thombergs**
This PR removes the global proxy configuration we've been using and replaces it with a more targeted proxy configuration on the level of the GitHub clients.

I have also added a special case so we're not using the proxy for our internal GitHub Server (which was the original reason to change this). Access to the GitHub Server has previously not worked, because the outbound proxy doesn't allow traffic back into the Atlassian network.

**End state:**
- all calls to GitHub APIs (both cloud and on-premise) go through the outbound proxy
- exception: calls to github.internal.atlassian.com don't use any proxy
- calls to the Jira API will no longer go through the outbound proxy, because I removed the global configuration.  This doesn't matter, however, since calls to Jira should never really have gone through the outbound proxy in the first place. I would have liked to keep the global proxy config behind a feature flag, but I'm not sure about how the global config and the local Axios config I added interfere.

This change is not feature flagged because I didn't find a good way to feature flag it. I tried a few things, but suggestions are welcome. I will test in staging if it still works with an IP allowlist (which is the only feature that might break due to proxy changes).

### CODE REVIEW

File: src/config/proxy.ts
Comment (thombergs) [inline src/config/proxy.ts:13]: This was the global proxy config that would be used for ALL http calls. We had to do this globally, because we couldn't teach Probot to use a specific proxy. Now that Probot is replaced with our own GitHub clients, we can use a more targeted proxy config.

File: src/config/proxy.ts
Comment (thombergs) [inline src/config/proxy.ts:16]: This Axios issue drove me nuts. Why is there a `proxy` option to configure a proxy host and port when it ALWAYS returns a 501?

**Comment by mboudreau** [comment]
@thombergs I think this needs to wait until we've completely removed probot/octokit.  @joshkay10 how's that going btw?  I think we're using the new clients everywhere, but still have some code for the old way.

File: src/github/client/github-app-client.ts
Comment (thombergs) [inline src/github/client/github-app-client.ts:40]: I moved the axios client out of the concrete GitHub clients and into the `GitHubClient` parent class. This was duplicate code and made it harder to configure the clients in a central place (for example to configure a proxy).

File: src/github/client/github-client-interceptors.ts
Comment (thombergs) [inline src/github/client/github-client-interceptors.ts:108]: At one point when I was investigating the 501 that Axios always returned I added some extra information to the error logs so I didn't have to guess which call went wrong.

File: src/github/client/github-client.ts
Comment (thombergs) [inline src/github/client/github-client.ts:49]: Moved the axios client to the `GitHubClient` parent class and added the proxy config (or not, depending on the `baseUrl`).

File: src/github/client/github-user-client.ts
Comment (thombergs) [inline src/github/client/github-user-client.ts:27]: This is using the parent class's `axios` instance now.

File: src/github/client/github-installation-client.ts
Comment (thombergs) [inline src/github/client/github-installation-client.ts:53]: This is using the parent class's `axios` instance now.

File: src/config/proxy.ts
Comment (mboudreau) [inline src/config/proxy.ts:12]: that thread at the end said to update to the latest version as it should be fixed now?

File: src/config/proxy.ts
Comment (mboudreau) [inline src/config/proxy.ts:10]: I don't like the idea that this util knows about AxiosRequestConfig data type.  

Plus, this is only being used in a single place right? do we even need this file at all? We could put all of it in `github-client`

File: src/github/client/github-client.ts
Comment (mboudreau) [inline src/github/client/github-client.ts:45]: I feel a great disturbance in the force...

should we be ignoring the proxy altogether for all `atlassian.com` or even `atlassian.net` urls so that it gets routed internally instead of going to the internet first?

**Comment by rachellerathbone** [comment]
> @thombergs I think this needs to wait until we've completely removed probot/octokit. @joshkay10 how's that going btw? I think we're using the new clients everywhere, but still have some code for the old way.

Probot? We're still in the process of implementing the new webhook receiver. We need this change so we can test in environments outside of our local so we can feel confident that all the GHE changes and the new webhook receiver work as expected. This needs to happen before we officially start ripping Probot out.

File: yarn.lock
Comment (mboudreau) [inline yarn.lock:3902]: this is wrong... it should be pointing to npmjs.org not PAC for this library

File: src/config/proxy.ts
Comment (mboudreau) [reply src/config/proxy.ts:16]: apparently, it has something to do with a combination of sub package node minimum TLS versions. This might be fixed in the latest version of Axios (0.27.2).  Maybe do an update while you're at it?

File: src/github/client/github-app-client.ts
Comment (mboudreau) [reply src/github/client/github-app-client.ts:40]: pretty sure you could move the interceptors with it as well to the superclass one

Comment (mboudreau) [reply src/github/client/github-app-client.ts:40]: while you're at it, can you also move the `get`, `post`, `delete`, `put` and `graphql` functions to the superclass?  fanx :)

File: yarn.lock
Comment (thombergs) [reply yarn.lock:3902]: Someday, I'll learn not to stumble over this stupid file....

**Comment by thombergs** [comment]
Tests in staging have been successful:

Enabled the IP allowlist in our org:
image

It does show the "blocked" error message when the IP range is not in the list:
image

Logs say this:
```
Although you appear to have the correct authorization credentials, the `Fusion-Arc` organization has an IP allow list enabled, and 13.52.5.107 is not permitted to access this resource.
```
which means that the calls have been made through the proxy.

And it doesn't show the error message when the IP range IS in the list:
image

File: src/github/client/github-app-client.ts
Comment (thombergs) [reply src/github/client/github-app-client.ts:40]: I'm happy to do that (and some other cleanup around the GitHub clients that I think should be done), but I'd prefer to do that in a second PR.

File: src/config/proxy.ts
Comment (thombergs) [reply src/config/proxy.ts:12]: I'll try that.

Comment (thombergs) [reply src/config/proxy.ts:12]: The comment "Please could you retry with the latest version and open a new issue should this error still be relevant?" combined with the status "possible bug" don't spark confidence in me, though :)

Comment (thombergs) [reply src/config/proxy.ts:12]: Nope, not working:
image

File: src/config/proxy.ts
Comment (thombergs) [reply src/config/proxy.ts]: moved the proxy config into the GitHubClient

File: src/github/client/github-client.ts
Comment (thombergs) [reply src/github/client/github-client.ts]: made it so that it matches everything that ends with atlassian.com, now

**Comment by joshkay10** [comment]
> > @thombergs I think this needs to wait until we've completely removed probot/octokit. @joshkay10 how's that going btw? I think we're using the new clients everywhere, but still have some code for the old way.
> 
> Probot? We're still in the process of implementing the new webhook receiver. We need this change so we can test in environments outside of our local so we can feel confident that all the GHE changes and the new webhook receiver work as expected. This needs to happen before we officially start ripping Probot out.

@mboudreau @rachellerathbone @thombergs  Everything is behind a feature flag and no Probot should have been called in at least a month.... however, I spend today pulling them all out so we can be sure!

File: src/github/client/github-client.ts
Comment (thombergs) [reply src/github/client/github-client.ts]: Added the tests now. Wanted to get some feedback on the PR before locking it in with tests.

**Comment by thombergs** [comment]
Alright, thanks for the update @joshkay10.

That means we should probably merge this only once the feature flags have been removed, i.e. when these PRs have been merged: 

- https://github.com/atlassian/github-for-jira/pull/1394
- https://github.com/atlassian/github-for-jira/pull/1393
- https://github.com/atlassian/github-for-jira/pull/1392
- https://github.com/atlassian/github-for-jira/pull/1391
- https://github.com/atlassian/github-for-jira/pull/1390
- https://github.com/atlassian/github-for-jira/pull/1389

File: src/github/client/github-client.ts
Comment (gxueatlassian) [inline src/github/client/github-client.ts:43]: /nit
Happy to leave as it is, just minor suggestion.

-------------

I know it is tricky to test, and therefore change from `private` to `public`.

In this case, I'd usually do bellow in the test file, so instead of testing private method, assert on the dependencies. 
```javascript
import axios from "axios";
it('should ...', async ()=>{
   jest.spyOn(axios, 'create');
   ... do stuff
   expect(axios.get).toBeCalledWith(expect.objectContaining({
          proxy: { agent: blah blah } 
   });
});
```

Another option will be using nock to assert the http call is going through proxy I guess, but seems much more code.

**Review by gxueatlassian — APPROVED** [review]

**Comment by gxueatlassian** [comment]
Approved, but please wait for more approvals.

File: src/github/client/github-client.ts
Comment (mboudreau) [inline src/github/client/github-client.ts:45]: I'm not sure this is going to work as you expect it to.  GHE needs to have `/api/` in baseUrl. Instead, I'd probably do something like `new URL(baseUrl).host.endsWith("atlassian.com")` .  Do we also want to include atlassian.net?

File: src/github/client/github-client.ts
Comment (mboudreau) [inline src/github/client/github-client.ts:48]: don't need the else if you're returning in the if

**Review by mboudreau — CHANGES_REQUESTED** [review]
just one small thing to change for the baseUrl so that it works with GHE

File: src/github/client/github-client.ts
Comment (thombergs) [reply src/github/client/github-client.ts]: good catch. Although, it worked in my tests. 

I made it more robust now ... just atlassian.com for now because I hope this is the only case we'll ever need (before we rip it out again)

File: src/github/client/github-client.ts
Comment (thombergs) [reply src/github/client/github-client.ts]: ✅

File: src/github/client/github-client.ts
Comment (thombergs) [reply src/github/client/github-client.ts:43]: It doesn't work with Nock, I had researched that. 

I had thought about mocking the axios client, but the current test is so much easier to read and maintain so I opted for that (even though it doesn't test the ultimate axios call).

**Review by mboudreau — APPROVED** [review]

**Pull request merged**
